tex2pdf_service/tex2pdf/doc_converter.py

```diff
@@ -45,46 +45,60 @@ def combine_documents(doc_list: typing.List[str], out_dir: str, out_filename: st
                     converted_docs.append(doc_path)
-                except PdfError:
-                    failed_docs.append(doc_path)
-                    logger.warning("Cannot open PDF file %s", doc_path, extra=log_extra)
-                    pass
-            elif ext.lower() in graphics_exts:
-                temp_pdf = os.path.join(out_dir, stem + '.pdf')
-                try:
-                    pdf_filename = convert_image_to_pdf(doc_path, temp_pdf)
-                    if pdf_filename and os.path.exists(pdf_filename):
-                        converted_docs.append(doc_path)
-                        with pikepdf.Pdf.open(pdf_filename) as pdf_page:
-                            pdf.pages.extend(pdf_page.pages)
-                        pass
-                except UnidentifiedImageError:
-                    failed_docs.append(doc_path)
-                    logger.warning("Unsupported %s", doc_path, extra=log_extra)
-                    pass
-                except Exception as _exc:
-                    failed_docs.append(doc_path)
-                    logger.warning("Unknown error %s", doc_path, extra=log_extra,
-                                   exc_info=True)
-                    pass
-                pass
-            pass
-        pdf.save(output_path)
-    return out_filename, strip_to_basename(converted_docs), strip_to_basename(failed_docs)
+            except UnidentifiedImageError:
+                failed_docs.append(doc_path)
+                logger.warning("Unsupported %s", doc_path, extra=log_extra)
+            except Exception as _exc:
+                failed_docs.append(doc_path)
+                logger.warning("Unknown error %s", doc_path, extra=log_extra,
+                               exc_info=True)
+    if len(effective_pdf_list) == 0:
+        # this should not happen, we should have at least one pdf from the tex file
+        raise Exception("No documents have been found.")
+    if len(effective_pdf_list) == 1:
+        # we didn't return above where we check for len(doc_list) == 1
+        # so this is an image that has been converted to pdf.
+        # Somehow surprising ... do we have such submissions?
+        if effective_pdf_list[0] != output_path:
+            shutil.move(effective_pdf_list[0], output_path)
+        converted_docs.append(os.path.basename(effective_pdf_list[0]))
+        return out_filename, strip_to_basename(converted_docs), strip_to_basename(failed_docs), addon_outcome
+        
+    # call gs to combine the pdf
+    # we cannot use pikepdf (easily) here since it breaks annotations (links)
+    gs_cmd = [
+        "gs", "-sDEVICE=pdfwrite", "-dNOPAUSE", "-dBATCH", "-dSAFER", f"-sOutputFile={output_path}"
+    ] + effective_pdf_list
+    logger.debug("Running gs to combine pdfs: %s", shlex.join(gs_cmd), extra=log_extra)
+    # exception handing is done in convert_driver:_finalize_pdf
+    ret = subprocess.run(gs_cmd, capture_output=True, timeout=30, check=True, text=True)
```

Comment:
If an exception is raised here, the converter driver generate_pdf will deal with it.